### PR TITLE
feat(cdn): get item log lines in json

### PR DESCRIPTION
**Pull request by richardlt**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14412.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘  
  * Standalone tests 1 -   
  * Standalone tests 2 -   
  * Standalone tests 3 -

**Review by yesnault — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14425.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/cdn/redis/reader.go [hunk 1]
```diff
@@ -2,29 +2,70 @@ package redis
...
 import (
 	"io"
-	"strings"
+)
+
+type ReaderFormat string
+
+const (
+	ReaderFormatJSON ReaderFormat = "json"
+	ReaderFormatText ReaderFormat = "text"
```
Comment (fsamin) [inline engine/cdn/redis/reader.go:11]: move in the parent package

**Review by fsamin — CHANGES_REQUESTED** [review]

**Review by sguiheux — APPROVED** [review]

File: engine/cdn/redis/reader.go (hunk 1 above)
Comment (richardlt) [reply engine/cdn/redis/reader.go]: moved to SDK

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14432.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14436.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Review by fsamin — COMMENTED** [review]
i think we have to check what we want to do about URLs and Authentication.

Do we want to expose URL for all CDN items for download ? Like artifacts in CDS ?
Do we need to set an expiration on this URL (like temp-url in swift)

File: engine/cdn/item_logs_handler.go [hunk 2]
```diff
@@ -66,27 +67,52 @@ func (s *Service) getItemLogsHandler() service.Handler {
...
-		// Check Authorization header
-		var token sdk.CDNAuthToken
-		v := authentication.NewVerifier(s.ParsedAPIPublicKey)
-		if err := v.VerifyJWS(tokenRaw, &token); err != nil {
-			return sdk.NewErrorWithStack(err, sdk.ErrUnauthorized)
+		return service.Write(w, rc, http.StatusOK, "application/json")
+	}
+}
+
+func (s *Service) getItemLogsDownloadHandler() service.Handler {
```
Comment (fsamin) [inline engine/cdn/item_logs_handler.go:103]: I think we should have a more generic handler that handles CDN item download whatever their type

File: engine/cdn/item_logs_handler.go [hunk 3]
```diff
@@ -120,3 +146,21 @@ func (s *Service) getSizeByProjectHandler() service.Handler {
 		return service.WriteJSON(w, size, http.StatusOK)
 	}
 }
+
+func (s *Service) checkItemLogsAuth(r *http.Request) (sdk.CDNAuthToken, error) {
+	vars := mux.Vars(r)
+	apiRef := vars["apiRef"]
+	tokenRaw := strings.TrimPrefix(r.Header.Get("Authorization"), "Bearer ")
```
Comment (fsamin) [inline engine/cdn/item_logs_handler.go:153]: not sure about that the apiRef should be in URL and in the JWT, i guess the JWT only is mandatory

File: engine/cdn/redis/readwrite.go [hunk 4]
```diff
@@ -17,16 +19,48 @@ type ReadWrite struct {
...
+	res := make([]string, to-from+1)
 	if err := rw.Store.ScoredSetScan(context.Background(), cache.Key(rw.PrefixKey, rw.ItemID), float64(from), float64(to), &res); err != nil {
-		return res, err
+		return nil, err
 	}
+	ls := make([]Line, len(res))
 	for i := range res {
-		res[i] = strings.TrimFunc(res[i], unicode.IsNumber)
-		res[i] = strings.TrimPrefix(res[i], "#")
+		tmp := strings.SplitN(res[i], "#", 2)
```
Comment (fsamin) [inline engine/cdn/redis/readwrite.go:51]: why did you change this ?
the previous implementation wasn't error prone.
Here is there is any weird data in redis, we are unable to read the whole log :(

**Review by fsamin — CHANGES_REQUESTED** [review]

File: engine/cdn/item_logs_handler.go (hunk 3 above)
Comment (richardlt) [reply engine/cdn/item_logs_handler.go:153]: Ok, Auth change will be in another PR.

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14441.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/cdn/redis/readwrite.go (hunk 4 above)
Comment (richardlt) [reply engine/cdn/redis/readwrite.go]: This was changed because we need to get the line number from redis, I edit my first version to get the value from redis score instead of parsing the log value.

File: engine/cdn/item_logs_handler.go (hunk 2 above)
Comment (richardlt) [reply engine/cdn/item_logs_handler.go]: I made some changes in download handler its more generic now. Filename is now human readable and generated from  values in the api ref.

**Review by fsamin — APPROVED** [review]
